File: tier2_intermediate/dynamic_programming/problems/p466_count_the_repetitions.mjs
/**
 * P466: Count The Repetitions [PREMIUM] (Hard)
 * https://leetcode.com/problems/count-the-repetitions/
 * Topics: String, Dynamic Programming
 *
 * We define str = [s, n] as the string str which consists of the string s concatenated n times.
 * We define that string s1 can be obtained from string s2 if we can remove some characters from s2 such that it becomes s1.
 * You are given two strings s1 and s2 and two integers n1 and n2. You have the two strings str1 = [s1, n1] and str2 = [s2, n2].
 * Return the maximum integer m such that str = [str2, m] can be obtained from str1.
 * Example 1:
 *     Input: s1 = "acb", n1 = 4, s2 = "ab", n2 = 2
 *     Output: 2
 *
 * Example 2:
 *     Input: s1 = "acb", n1 = 1, s2 = "acb", n2 = 1
 *     Output: 1
 *
 * Constraints:
 *     - 1 <= s1.length, s2.length <= 100
 *     - s1 and s2 consist of lowercase English letters.
 *     - 1 <= n1, n2 <= 106
 *
 * Template (python3):
 *     class Solution:
 *         def getMaxRepetitions(self, s1: str, n1: int, s2: str, n2: int) -> int:
 */

function solve(s1, n1, s2, n2) {
  throw new Error("NotImplementedError");
}

const tests = [
  { label: "example 1", input: ["acb", 4, "ab", 2], expected: 2 },
  { label: "example 2", input: ["acb", 1, "acb", 1], expected: 1 },
  { label: "single char repeated", input: ["a", 100, "a", 1], expected: 100 },
  { label: "impossible char", input: ["a", 1, "b", 1], expected: 0 },
  { label: "each s1 yields one s2 match", input: ["abc", 10, "ac", 1], expected: 10 },
  { label: "overlap matching", input: ["aba", 3, "ab", 1], expected: 3 },
];
let passed = 0;
for (let i = 0; i < tests.length; i++) {
  const t = tests[i];
  const got = solve(...t.input);
  if (JSON.stringify(got) === JSON.stringify(t.expected)) {
    passed++;
    console.log(`  Test ${i + 1} (${t.label}): PASS`);
  } else {
    console.log(`  Test ${i + 1} (${t.label}): FAIL`);
    console.log(`    Expected: ${JSON.stringify(t.expected)}\n    Got:      ${JSON.stringify(got)}`);
  }
}
console.log(`\n  ${passed}/${tests.length} passed`);
process.exit(passed === tests.length ? 0 : 1);

File: tier2_intermediate/dynamic_programming/test_runners/p466_count_the_repetitions.py
from src.utils.judge_base import JudgeBase, TestCase, register


class P466CountTheRepetitions(JudgeBase):
    name = "466. Count The Repetitions"
    test_cases = [
        TestCase(input=("acb", 4, "ab", 2), expected=2, label="example 1"),
        TestCase(input=("acb", 1, "acb", 1), expected=1, label="example 2"),
        TestCase(input=("a", 100, "a", 1), expected=100, label="single char repeated"),
        TestCase(input=("a", 1, "b", 1), expected=0, label="impossible char"),
        TestCase(input=("abc", 10, "ac", 1), expected=10, label="each s1 yields one s2 match"),
        TestCase(input=("aba", 3, "ab", 1), expected=3, label="overlap matching"),
    ]

    def to_stdin(self, tc_input) -> str:
        s1, n1, s2, n2 = tc_input
        return self.serialize_multi(
            self.serialize_string(s1),
            self.serialize_int(n1),
            self.serialize_string(s2),
            self.serialize_int(n2),
        )

    def check_stdout(self, stdout: str, tc_expected) -> bool:
        return self.parse_int(stdout) == tc_expected


register(P466CountTheRepetitions)
